XF Upgrade 4.0

SergejDK opened the pull request:
fix #412 
fix #417 

Fuget: [https://www.fuget.org/packages/Xamarin.Forms/4.0.0.425677/lib/netstandard2.0/diff/3.6.0.344457/](https://www.fuget.org/packages/Xamarin.Forms/4.0.0.425677/lib/netstandard2.0/diff/3.6.0.344457/)
There are some open parts on this PR:

- [x] Navigationpage
-- TitleIconImageSource is the only part: 
@TimLariviere  is it an attachedproperty?

- [x] Searchhandler: selecteditem
- [x] shellnavigationstate: location

TimLariviere commented on tools/Generator/Xamarin.Forms.Core.json:
```diff
@@ -2261,6 +2292,69 @@
+          "defaultValue": "false",
+          "modelType": "bool",
+          "inputType": "bool",
+          "updateCode": "updateIsFocusesd"
```
Little typo here :)

SergejDK replied:
True. Maybe not such a good idea to work on this in the night :D

TimLariviere commented:
> Navigationpage
-- TitleIconImageSource is the only part:
@TimLariviere is it an attachedproperty?

It is. Just like the now obsoleted TitleIcon.
https://github.com/xamarin/Xamarin.Forms/blob/a19e105240495e7a58980ca205cbe22ae85a7f0d/Xamarin.Forms.Core/NavigationPage.cs#L29

SergejDK commented:
@TimLariviere 
do you think we need the ShellNavigationState?
For me it seems like the class is mostly used internally and not sth. that we should provide?
This was in the upgrade to 3.5 but I think we do not need it.

TimLariviere commented:
>do you think we need the ShellNavigationState?
>For me it seems like the class is mostly used internally and not sth. that we should provide?

You're right, we shouldn't provide it.
It's a read-only class built by Xamarin.Forms (only 1 get only property)
https://docs.microsoft.com/en-us/dotnet/api/xamarin.forms.shellnavigationstate.location?view=xamarin-forms

According to the documentation:
>The ShellNavigationState object is constructed by the GoToAsync method, from a string, or a Uri, and it has its Location property set to the string or Uri argument.

https://docs.microsoft.com/en-us/xamarin/xamarin-forms/app-fundamentals/shell/navigation#perform-navigation

SergejDK commented:
Well then I think everything should be in place and ready for merge.
I think travis does not want to work today ....

TimLariviere commented:
Travis doesn't seem very happy with the code (for once it's not Xamarin.Android)

>/Users/travis/build/fsprojects/Fabulous/samples/AllControls/AllControls/AllControls.fs(381,78): error FS0495: The member or object constructor 'Label' has no argument or settable return property 'styleClass'. The required signature is static member View.Label : ?text:string * ?horizontalTextAlignment:TextAlignment * ?verticalTextAlignment:TextAlignment * ?fontSize:obj * ?fontFamily:string * ?fontAttributes:FontAttributes * ?textColor:Color * ?formattedText:ViewElement * ?lineBreakMode:LineBreakMode * ?lineHeight:double * ?maxLines:int * ?textDecorations:TextDecorations * ?horizontalOptions:LayoutOptions * ?verticalOptions:LayoutOptions * ?margin:obj * ?gestureRecognizers:ViewElement list * ?anchorX:double * ?anchorY:double * ?backgroundColor:Color * ?heightRequest:double * ?inputTransparent:bool * ?isEnabled:bool * ?isVisible:bool * ?minimumHeightRequest:double * ?minimumWidthRequest:double * ?opacity:double * ?rotation:double * ?rotationX:double * ?rotationY:double * ?scale:double * ?translationX:double * ?translationY:double * ?widthRequest:double * ?resources:(string * obj) list * ?styles:Style list * ?styleSheets:StyleSheets.StyleSheet list * ?isTabStop:bool * ?scaleX:double * ?scaleY:double * ?tabIndex:int * ?childrenReordered:(EventArgs -> unit) * ?measureInvalidated:(EventArgs -> unit) * ?focused:(FocusEventArgs -> unit) * ?sizeChanged:(Fabulous.CustomControls.SizeChangedEventArgs -> unit) * ?unfocused:(FocusEventArgs -> unit) * ?visual:IVisual * ?classId:string * ?styleId:string * ?automationId:string * ?created:(Label -> unit) * ?ref:ViewRef<Label> -> ViewElement.

TimLariviere commented on tools/Generator/Xamarin.Forms.Core.json:
```diff
@@ -95,18 +95,6 @@
           "defaultValue": "1.0"
         },
-        {
-          "name": "Style",
```
Is it supposed to be removed?

TimLariviere replied:
Along with StyleClass

SergejDK replied:
I took it from fuget:
```fsharp
    public IList<string> class { get; set; }
    public Style Style { get; set; }
    public IList<string> StyleClass { get; set; }
```
In the Visualelement it looks like it does not exist anylonger

TimLariviere replied:
It has been moved to NavigableElement now, which VisualElement inherits
https://docs.microsoft.com/en-us/dotnet/api/xamarin.forms.navigableelement?view=xamarin-forms

TimLariviere replied:
Comparing 3.4.0 vs 4.0.0, fuget shows that NavigableElement was created with these 3 properties
https://www.fuget.org/packages/Xamarin.Forms/4.0.0.425677/lib/netstandard2.0/diff/3.4.0.1008975/

TimLariviere replied:
Maybe we will need to change the Generator to prevent exposing those abstract intermediate classes, though.
Otherwise we will polute the View.* with non-usable controls

SergejDK replied:
will change it. didn't see that it was moved.

SergejDK replied:
That's right that we should change the generator a bit. Was mentioned in a issue, too. But that will be another PR... First of all the upgrade to 4.0

SergejDK commented:
The `SearchBoxVisibility`-Problem happens in the creation of the testnuget part.
I think that the problem here is that in the last version of XF they had a typo in it and now they corrected it. But the templates are using the "old" Fabulous version 0.34 which is using the `SearchBoxVisiblity` with the typo. 
I will try to reference my local version.

TimLariviere commented:
`./build.sh Test` should generate a local nupkg for Fabulous and use it to run the samples.
So you shouldn't have a problem with SearchBoxVisibility when running this command.

SergejDK commented:
The interesting part is that I get this error. Will push my latest changes in the next hour.

SergejDK commented:
@TimLariviere  
kinda works here. Maybe I changed sth. locally....

TimLariviere commented:
I also have the error with `SearchBoxVisiblity`.
I suspect it's because we're still using the same version number as present on nuget.org.
When the testapp is trying to restore a `Fabulous v0.34.0`, it is already in cache and NuGet restore won't search for a local package.

The easiest way to avoid that error is being temporarily changing the version number in the README.md

TimLariviere commented:
@SergejDK I can't find `BaseShellItem` in the json file. Is it missing?
It has some properties like `IsTabStop` and `TabIndex`
https://docs.microsoft.com/en-us/xamarin/xamarin-forms/release-notes/4.0/4.0.0-api#type-changed-xamarinformsbaseshellitem

SergejDK commented:
@TimLariviere 
Honestly I don't know. Maybe i included it in 3.5 and deleted in 3.6 but don't know why. Currently on my phone and can't check easy where I throw it away. 
Will include it. 

For the searchboxvisibility problem we could write sth. In the docs for the time being. After a new release it should work just fine ?

TimLariviere commented:
>For the searchboxvisibility problem we could write sth. In the docs for the time being. After a new release it should work just fine ?

Yes. That would be a good idea.

SergejDK commented:
@TimLariviere 
added BaseShellItem and changed to EventArgs. I hope that will work fine. Didn't get the time to check it locally. 

For the docs - could you write sth. there, please? -  I'm out of town till sunday. Would be great so we can maybe merge it if everything works

TimLariviere commented:
All's good.
Merging now :)

xperiandri commented:
@TimLariviere, maybe you would better bump version to 0.40 to indicate that it has some significant changes?

xperiandri commented:
@SergejDK, works perfectly. Thank you very much for a great instant update.

TimLariviere commented:
@xperiandri In my opinion, even with this major update to support Xamarin.Forms 4.0, Fabulous remains mostly the same. There's only a bunch of new controls supported, which most updates of Fabulous add.
